### PR TITLE
[dataset] support bucket by seq length

### PR DESCRIPTION
**Pull request by Mddct**
follow this pr https://github.com/wenet-e2e/wenet/pull/2316

TODO:
- [x] unit test
- [x] aishell training

### CODE REVIEW

**Comment by Mddct** [comment]
it works

Screenshot 2024-02-02 at 00 57 11

**Comment by Mddct** [comment]
尝试跑了下更大的bucket 的 82->128, 训练时间 14h10min （和上边一致）， 性能几乎一致

```yaml
    batch_conf:
        batch_type: 'bucket' # static or bucket or dynamic
        bucket_boundaries:  [500, 1000, 1500]
        bucket_batch_sizes: [128, 64, 32, 16]
```
Screenshot 2024-02-03 at 17 12 22

**Comment by Mddct** [comment]
以下增大bucket的测试：
```yaml
    batch_conf:
        batch_type: 'bucket' # static or bucket or dynamic
        bucket_boundaries:  [500, 1000, 1500]
        bucket_batch_sizes: [128, 64, 32, 16]
```

- raw模式， 训练时见14h11min
Screenshot 2024-02-04 at 09 45 21

- shard 模式， 训练时间13h10min

Screenshot 2024-02-04 at 09 43 26

**Review by xingchensong — APPROVED** [review]

File: wenet/dataset/datapipes.py [hunk 1]
```diff
@@ -56,6 +58,94 @@ def __iter__(self):
...
+                result = self.wrapper_class(self.buffer_elements[key])
+                yield result
+                self.curr_buffer_size -= len(self.buffer_elements[key])
+                del self.buffer_elements[key]
+
+            if self.curr_buffer_size == self.max_buffer_size:
+                result_to_yield = self._remove_biggest_key()
+                if result_to_yield is not None:
+                    result = self.wrapper_class(result_to_yield)
+                    yield result
```
Comment (xingchensong) [inline wenet/dataset/datapipes.py:141]: why remove biggest_key？

Comment (Mddct) [reply wenet/dataset/datapipes.py:141]: 防止长时间桶内数据不够拼batch 不产生数据， 

目前这个几乎不会发生， 从torch里边的group 实现copy过来的

**Comment by Mddct** [comment]
这里列下上边两个表格中提到的一些指标， 直观一些
|batch size      |data type |  训练时间 |  att/rescore/ctc greedy/ctc beam wer|
| ----------- | ----------- |----------- |----------- |
|     bucket_boundaries:  [500, 1000, 1500]  <br>  bucket_batch_sizes: [82, 64, 32, 16]| raw       |  14h13min| 5.58/5.27/5.82/5.82|
|     bucket_boundaries:  [500, 1000, 1500]  <br>  bucket_batch_sizes: [128, 64, 32, 16]| raw       | 14h11min |5.53/5.24/5.85/5.85|
|     bucket_boundaries:  [500, 1000, 1500]  <br>  bucket_batch_sizes: [128, 64, 32, 16] |  shard     | 13h 10min|  5.59/5.20/5.73/5.72|

**Review by robin1001 — APPROVED** [review]

**Pull request merged**